src/Control/Newtype/Generic.hs

```diff
@@ -0,0 +1,203 @@
+{-# language CPP                        #-}
+{-# language FlexibleInstances          #-}
+{-# language MultiParamTypeClasses      #-}
+{-# language ScopedTypeVariables        #-}
+{-# language TypeFamilies               #-}
+{-# language TypeOperators              #-}
+{-# language DataKinds                  #-}
+{-# language UndecidableInstances       #-}
+
+{- |
+A version of the 'Newtype' typeclass and related functions.
+Primarily pulled from Conor McBride's Epigram work. Some examples:
+
+>>> ala Sum foldMap [1,2,3,4]
+10
+
+>>> ala Endo foldMap [(+1), (+2), (subtract 1), (*2)] 3
+8
+
+>>> under2 Min (<>) 2 1
+1
+
+>>> over All not (All False)
+All {getAll = True)
+
+The version of the 'Newtype' class exported by this module has only
+one instance, which cannot be overridden. It has instances for
+*all* and *only* newtypes with 'Generic' instances whose generated
+'Coercible' instances are visible. Users need not, and cannot, write
+their own instances.
+
+Like McBride's version, and unlike the one in @newtype-generics@,
+this version has two parameters: one for the newtype and one for the
+underlying type. This generally makes 'Newtype' constraints more compact
+and easier to read.
+
+Unlike the versions in "CoercibleUtils", one of the "packers" must
+operate on only one newtype layer at a time; this improves inference at
+the expense of some flexibility.
+
+Note: Each function in this module takes an argument representing a
+newtype constructor. This is used only for its type. To make that clear,
+the type of that argument is allowed to be extremely polymorphic: @o \`to\` n@
+rather than @o -> n@.
+
+@since TODO
+-}
+module Control.Newtype.Generic
+  ( Newtype
+  , O
+  , op
+  , ala
+  , ala'
+  , under
+  , over
+  , under2
+  , over2
+  , underF
+  , overF
+  ) where
+
+import GHC.Generics
+import Data.Coerce
+#if MIN_VERSION_base(4,9,0)
+import GHC.TypeLits (TypeError, ErrorMessage (..))
+#endif
+import CoercibleUtils (op)
+
+-- | Get the underlying type of a newtype.
+--
+-- @
+-- data N = N Int deriving Generic
+-- -- O N = Int
+-- @
+type O x = GO x (Rep x)
+
+type family GO x rep where
+#if __GLASGOW_HASKELL__ >= 800
+  GO _x (D1 ('MetaData _n _m _p 'True) (C1 _c (S1 _s (K1 _i a)))) = a
+  GO x _rep = TypeError
+    ('Text "There is no " ':<>: 'ShowType Newtype ':<>: 'Text " instance for"
+      ':$$: 'Text "    " ':<>: 'ShowType x
+      ':$$: 'Text "because it is not a newtype.")
+#else
+  GO _x (D1 _d (C1 _c (S1 _s (K1 _i a)))) = a
+#endif
+
+-- | @Newtype n o@ means that @n@ is a newtype wrapper around
+-- @o@. @n@ must be an instance of 'Generic'. Furthermore, the
+-- @'Coercible' n o@ instance must be visible; this typically
+-- means the newtype constructor is visible, but the instance
+-- could also have been brought into view by pattern matching
+-- on a 'Data.Type.Coercion.Coercion'.
+class (Coercible n o, O n ~ o) => Newtype n o
```

Comment:
All my instincts say that this should be `Newtype o n`, because `Newtype o` is a potentially useful partial application. I've stuck with this order because it's the one used in the `newtype` package, and that compatibility makes it easier to make the switch.